hazelcast/src/main/java/com/hazelcast/cache/impl/CachePartitionsIterator.java

```diff
@@ -70,27 +70,24 @@ protected List fetch() {
         final OperationService operationService = cacheProxy.getNodeEngine().getOperationService();
         if (prefetchValues) {
             Operation operation = cacheProxy.operationProvider.createFetchEntriesOperation(pointers, fetchSize);
-            InternalCompletableFuture<CacheEntriesWithCursor> f = operationService
-                    .invokeOnPartition(CacheService.SERVICE_NAME, operation, partitionIndex);
-            CacheEntriesWithCursor iteratorResult = f.joinInternal();
-            if (iteratorResult != null) {
-                setLastTableIndex(iteratorResult.getEntries(), iteratorResult.getPointers());
-                return iteratorResult.getEntries();
-            }
+            CacheEntriesWithCursor iteratorResult = invoke(operationService, operation);
+            setIterationPointers(iteratorResult.getEntries(), iteratorResult.getPointers());
+            return iteratorResult.getEntries();
         } else {
             Operation operation = cacheProxy.operationProvider.createFetchKeysOperation(pointers, fetchSize);
-            InternalCompletableFuture<CacheKeysWithCursor> f = operationService
-                    .invokeOnPartition(CacheService.SERVICE_NAME, operation, partitionIndex);
-            CacheKeysWithCursor iteratorResult = f.joinInternal();
-            if (iteratorResult != null) {
-                setLastTableIndex(iteratorResult.getKeys(), iteratorResult.getPointers());
-                return iteratorResult.getKeys();
-            }
+            CacheKeysWithCursor iteratorResult = invoke(operationService, operation);
+            setIterationPointers(iteratorResult.getKeys(), iteratorResult.getPointers());
```

Comment:
The way I see it, it boils down to the return values from `ICacheRecordStore#fetchKeys` and `ICacheRecordStore#fetchEntries` and I don't see `null` being returned from either. Where did you see the `null` value?

Reply:
`f.joinInternal()` -> `join()` -> `reportJoin()`. The `reportJoin()` returns `null` in some cases

```
    private static <T> T reportJoin(Object r) {
        if (r instanceof AltResult) {
            Throwable x;
            if ((x = ((AltResult)r).ex) == null)
                return null;
            if (x instanceof CancellationException)
                throw (CancellationException)x;
            if (x instanceof CompletionException)
                throw (CompletionException)x;
            throw new CompletionException(x);
        }
        @SuppressWarnings("unchecked") T t = (T) r;
        return t;
    }
```

Reply:
Hm, not sure in which case this might happens but we'll have to inspect all of the codebase then as I see we have 290 usages of `InternalCompletableFuture#joinInternal` and I haven't seen null checks anywhere.

Reply:
Apparently this is a very rare case. @vbekiaris, WDYT?

Reply:
I see one reason why `AltResult.ex` might be `null` is if the future is completed with `null`:
```
    /** The encoding of the null value. */
    static final AltResult NIL = new AltResult(null);
```

Reply:
`AltResult(ex = null)` is the JDK 's `null` outcome encoding:

```
CompletableFuture cf = new CompletableFuture();
cf.complete(null); // --> this is encoded internally as an AltResult with field ex == null
```

The `CompletableFuture`s obtained from invoking `fetchKeys/Entries` will either complete normally (with the non-null value returned from the respective operation) or exceptionally (when there is an invocation issue) but `null` is not a possible outcome.